### PR TITLE
Add low-rank-modified metric

### PR DESCRIPTION
**Pull request by aseyboldt**
# Add a low-rank-modified metric

This adds a new type of metric for hmc-style samlers:
Instead of just allowing a diagonal or a full mass matrix, we combine the diagonal mass matrix with a couple of additional eigenvectors and eigenvalues of the remaining correlation.

This leads to the following form for the mass matrix:

$$
\begin{align}
P &= V(\Sigma^{-1} - I)V^T + I \\
M &= D^{-\frac{1}{2}}PD^{-\frac{1}{2}}
\end{align}
$$

$P$ has exactly the eigenvalues specified in $\Sigma$ with the corresponding eigenvectors in $V \in R^{n, k}$. All remaining eigenvalues are one.

All operations needed in hmc/nuts can be done in $O(nk)$, so if the number of eigenvalues is small this has a cost much closer to diagonal mass matrix estimation.

This PR for now only contributes the metric, but corresponding adaptation should come a bit later...

Related issue: #683

 A few important guidelines and requirements before we can merge your PR:

 - [ ] *If I add a new sampler*, there is an issue discussing it already;
 - [ ] We should be able to understand what the PR does from its title only;
 - [x] There is a high-level description of the changes;
 - [x] There are links to *all* the relevant issues, discussions and PRs;
 - [x] The branch is rebased on the latest `main` commit;
 - [x] Commit messages follow these [guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html);
 - [ ] The code respects the current naming conventions;
 - [ ] Docstrings follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html)
 - [x] `pre-commit` is installed and configured on your machine, and you ran it before opening the PR;
 - [ ] There are tests covering the changes;
 - [ ] The doc is up-to-date;
 - [ ] *If I add a new sampler** I added/updated related [examples](https://github.com/blackjax-devs/blackjax/tree/main/examples)

Consider opening a **Draft PR** if your work is still in progress but you would like some feedback from other contributors.

### CODE REVIEW

File: blackjax/mcmc/metrics.py [hunk 1]
```diff
@@ -208,6 +213,120 @@ def is_turning(
...
+        rho = m_sum - (m_right + m_left) / 2
+        turning_at_left = jnp.dot(velocity_left, rho) <= 0
+        turning_at_right = jnp.dot(velocity_right, rho) <= 0
+        return turning_at_left | turning_at_right
+
+    return Metric(
+        momentum_generator,
+        kinetic_energy,
+        is_turning,
+        data=(diagonal_scale_std, eigenvalues, eigenvectors),
```
Comment (junpenglao) [inline blackjax/mcmc/metrics.py:326]: I guess you want to store these as properties so it is easier for tuning later on? In any case I suggest removing it here, and add it in the subsequent PR when tuning is introduced (so we can discuss whether it is necessary)

Comment (aseyboldt) [reply blackjax/mcmc/metrics.py:326]: Yes, that was mostly for debugging and I forgot to take it out. Sorry :-)

File: blackjax/mcmc/metrics.py [hunk 2]
```diff
@@ -208,6 +213,120 @@ def is_turning(
...
 
+def gaussian_euclidean_low_rank(
+    diagonal_scale_std: Array,
+    eigenvectors: Array,
+    eigenvalues: Array,
+) -> Metric:
+    r"""Hamiltonian dynamic on euclidean manifold with normally-distributed momentum
+    :cite:p:`betancourt2013general`.
+
+    The gaussian euclidean metric is a euclidean metric further characterized
```
Comment (twiecki) [inline blackjax/mcmc/metrics.py:224]:
```suggestion
    The Gaussian Euclidean metric is a Euclidean metric further characterized
```

File: blackjax/mcmc/metrics.py [hunk 3]
```diff
@@ -208,6 +213,120 @@ def is_turning(
...
+    by setting the conditional probability density :math:`\pi(momentum|position)`
+    to follow a standard gaussian distribution. A Newtonian hamiltonian
```
Comment (twiecki) [inline blackjax/mcmc/metrics.py:226]:
```suggestion
    to follow a standard Gaussian distribution. A Newtonian Hamiltonian
```
